### ISSUE TITLE
ActiveSupport Notification support

### ISSUE THREAD

**Issue by teamon** [plataformatec/faraday-http-cache#20]
I'd like to display log information about external http requests in rails app. Right now I'm doing something like this:

```ruby
class FaradayInstrumentation < Faraday::Middleware
  dependency 'active_support/notifications'

  def call(env)
    ActiveSupport::Notifications.instrument('external.faraday', env) do
      @app.call(env)
    end
  end
end

MyApi.configure(...) do |c|
  c.use FaradayInstrumentation
  c.response :raise_error
end

```

and then

```ruby
ActiveSupport::Notifications.subscribe('external.faraday') do |name, starts, ends, _, env|
  url = env[:url]
  http_method = env[:method].to_s.upcase
  duration = (ends - starts)*1000
  Rails.logger.info '[%s] %s %s (%.3f ms)' % [url.host, http_method, url.request_uri, duration]
  $statsd.timing("external.http", duration, :tags => [
    "path:#{url.path}", "target:#{url.host}", "method:#{http_method}"
  ])
end
```

Besides statsd measurements it gives me line in logs like:

```
[some.host.com] GET /some/path (143.494 ms)
```

I know faraday-http-cache has `:logger` options, but that by default gives me two lines in logs. The prefect logs would be like:

```
[some.host.com] GET /some/path (fresh)
[some.host.com] GET /some/path (143.494 ms, miss)
```

etc. What's even more important is sending event to statsd, which in case of cache fresh could be just skipped. 

Ideally information about cache action/status would be put in `env` hash which is accessible in all middlewares (and in my instrumentation hook).

Any thoughts?

**Comment by dasch** [comment]
@lucasmazza I'd like to work on this – I need to instrument our use of the middleware using both extended logging and statsd, and I'd rather share the implementation back here instead of forking our own version.

After reading this thread I think there are two options:

1) Just add the trace to the env. This avoids adding any dependencies, and people who wish to instrument the cache can add their own middleware for doing so.
2) Emit a single AS::Notifications event after processing a request, containing the trace and the env – maybe the validation request env separately.

The first would of course be easier, but do you feel that there are important pieces of information that would be lost if only the trace is communicated?

**Comment by dasch** [comment]
I've implemented 1) in #63.

**Comment by lucasmazza** [comment]
Closed due #63 and #64, and version `1.1.0` is out :+1:
